TNL-4356 Multi-select on bulk email

efischer19 opened the pull request:
[Support multi-selection for bulk email recipients](https://openedx.atlassian.net/browse/TNL-4356)

efischer19 commented on common/djangoapps/util/models.py:
```diff
@@ -62,3 +63,49 @@ def to_python(self, value):
         return value
+
+
+class SerializedChoicesField(models.CharField):
```
I put this class in this file because it seems to be something that could be more generally useful than just the application I'm using in this PR.

efischer19 replied:
General documentation for custom fields is here, it describes the `to_python` and `get_prep_value` overrides.

https://docs.djangoproject.com/en/1.8/howto/custom-model-fields/

efischer19 commented on lms/djangoapps/bulk_email/migrations/0003_multiple_choices_to_option.py:
```diff
@@ -0,0 +1,20 @@
+            name='to_option',
+            field=util.models.SerializedChoicesField(default=[b'myself'], max_length=64, choices=[(b'myself', b'Myself'), (b'staff', b'Staff and instructors'), (b'learners', b'All students')]),
+        ),
+    ]
```
Because of how I designed `SerialChoicesField`, this migration becomes a nullop, and all previous data is still valid.
```
edxapp@precise64:~/edx-platform$ ./manage.py lms --settings=aws sqlmigrate bulk_email 0003
2016-04-29 15:54:39,215 INFO 15672 [dd.dogapi] dog_stats_api.py:66 - Initializing dog api to use statsd: localhost, 8125
System check identified some issues:

WARNINGS:
wiki.ArticleRevision.ip_address: (fields.W900) IPAddressField has been deprecated. Support for it (except in historical migrations) will be removed in Django 1.9.
	HINT: Use GenericIPAddressField instead.
```

cahrens commented on lms/djangoapps/bulk_email/migrations/0003_multiple_choices_to_option.py:
```diff
@@ -0,0 +1,20 @@
+        migrations.AlterField(
+            model_name='courseemail',
+            name='to_option',
+            field=util.models.SerializedChoicesField(default=[b'myself'], max_length=64, choices=[(b'myself', b'Myself'), (b'staff', b'Staff and instructors'), (b'learners', b'All students')]),
```
I worry about max_length only being 64 chars. We are plan to add a number of new options here (for cohort selection), and I don't think they will all fit.

Also, the new choices being added will need to be dynamic based on the cohorts in existence at the time. See https://openedx.atlassian.net/browse/TNL-4357.

efischer19 commented on lms/djangoapps/bulk_email/migrations/0003_add_emailtarget.py:
```diff
@@ -0,0 +1,37 @@
+            name='target',
+            field=models.ManyToManyField(to='bulk_email.EmailTarget'),
+        ),
+    ]
```
Results of sqlmigrate for this migration:
```
BEGIN;
CREATE TABLE `bulk_email_emailtarget` (`id` integer AUTO_INCREMENT NOT NULL PRIMARY KEY, `group_name` varchar(255) NOT NULL, `course_id` varchar(255) NOT NULL);
ALTER TABLE `bulk_email_emailtarget` ADD CONSTRAINT `bulk_email_emailtarget_group_name_4ab66d6f94c30656_uniq` UNIQUE (`group_name`, `course_id`);
CREATE TABLE `bulk_email_courseemail_target` (`id` integer AUTO_INCREMENT NOT NULL PRIMARY KEY, `courseemail_id` integer NOT NULL, `emailtarget_id` integer NOT NULL, UNIQUE (`courseemail_id`, `emailtarget_id`));
CREATE INDEX `bulk_email_emailtarget_ea134da7` ON `bulk_email_emailtarget` (`course_id`);
ALTER TABLE `bulk_email_courseemail_target` ADD CONSTRAINT `bul_courseemail_id_3c7523114edc8112_fk_bulk_email_courseemail_id` FOREIGN KEY (`courseemail_id`) REFERENCES `bulk_email_courseemail` (`id`);
ALTER TABLE `bulk_email_courseemail_target` ADD CONSTRAINT `bulk_emailtarget_id_fe911c9e68c00fc_fk_bulk_email_emailtarget_id` FOREIGN KEY (`emailtarget_id`) REFERENCES `bulk_email_emailtarget` (`id`);

COMMIT;
```

efischer19 commented:
So a big issue that's come to light is that the `CourseEmail` model currently has a `to_option` field, which is an enumeration of allowed values. Only a single option my be selected at a time. The new paradigm needs to have the ability for any number of arbitrarily valued options associated with a `CourseEmail`. To solve this, I am proposing the addition of a new `EmailTarget` model, which will form a `ManyToMany` relationship with `CourseEmail`.

The problem, as usual, will be "how can we migrate to this new approach without breaking things?" Here's what I'm thinking.
- Release 1:
  - schema migration: add new model table
  - code change: try to read new `EmailTarget` field, if it DNE fallback to old `to_option` field. Add logging for when this fallback occurs.
    - also fail gracefully if we get into this scenario and the old table has  been deleted
  - code change: write new `EmailTarget` field + old `to_option` field
    - will require restricting new to only use old values!
    - should not error as long as one of the 2 attempts succeeds
    - allows for full rollback if problems are detected
  - At this point, we're live in prod with the first part of the migration. The following data migrations can be run out-of-band in a background task.
    - data migration: seed new table with default values (`(course_id: None, group_name: SEND_TO_MYSELF)` for all currently possible `to_option` targets)
    - data migration: update all existing entries to the corresponding new
-Release 2:
  - before this, ensure that logging has indicated a successful switchover
    - if not fully successful, something is wrong and we need to reconsider
  - schema migration: drop old column
  - code change: remove “read with fallback” code, only read new `EmailTarget` field
  - code change: write only new `EmailTarget` field, remove "must match previous options" restriction

efischer19 commented:
fwiw, there are approximately 26k entries in this table on production currently
```
mysql> select count(*) from bulk_email_courseemail;
+----------+
| count(*) |
+----------+
|    26231 |
+----------+
1 row in set (0.04 sec)
```

efischer19 commented:
Data migration takes 1'42" on a comparable sandbox environment test. So we'll need to ensure the query is set up so as to only use a row lock at a time, not a table lock.
```
edxapp@efischer19:~/edx-platform$ time ./manage.py lms --settings=aws migrate bulk_email
Operations to perform:
  Apply all migrations: bulk_email
Running migrations:
  Rendering model states... DONE
  Applying bulk_email.0003_add_emailtarget... OK
  Applying bulk_email.0004_emailtarget_data... OK

real	1m42.117s
user	1m22.881s
sys	0m5.172s
```

efischer19 commented on lms/djangoapps/bulk_email/migrations/0003_auto_20160503_1218.py:
```diff
@@ -0,0 +1,76 @@
+            name='targets',
+            field=models.ManyToManyField(to='bulk_email.Target'),
+        ),
+    ]
```
Output of sqlmigrate:
```
edxapp@precise64:~/edx-platform$ ./manage.py lms --settings=devstack sqlmigrate bulk_email 0003
BEGIN;
CREATE TABLE `bulk_email_target` (`id` integer AUTO_INCREMENT NOT NULL PRIMARY KEY, `target_type` varchar(64) NOT NULL);
CREATE TABLE `bulk_email_cohorttarget` (`target_ptr_id` integer NOT NULL PRIMARY KEY, `cohort_id` integer NOT NULL);
CREATE TABLE `bulk_email_courseemail_targets` (`id` integer AUTO_INCREMENT NOT NULL PRIMARY KEY, `courseemail_id` integer NOT NULL, `target_id` integer NOT NULL, UNIQUE (`courseemail_id`, `target_id`));
ALTER TABLE `bulk_email_cohorttarget` ADD CONSTRAINT `bulk_emai_target_ptr_id_7974c77c83c2899d_fk_bulk_email_target_id` FOREIGN KEY (`target_ptr_id`) REFERENCES `bulk_email_target` (`id`);
ALTER TABLE `bulk_email_cohorttarget` ADD CONSTRAINT `b_cohort_id_3d66a5e8e283dba0_fk_course_groups_courseusergroup_id` FOREIGN KEY (`cohort_id`) REFERENCES `course_groups_courseusergroup` (`id`);
ALTER TABLE `bulk_email_courseemail_targets` ADD CONSTRAINT `bul_courseemail_id_47818d2b9b38e0e0_fk_bulk_email_courseemail_id` FOREIGN KEY (`courseemail_id`) REFERENCES `bulk_email_courseemail` (`id`);
ALTER TABLE `bulk_email_courseemail_targets` ADD CONSTRAINT `bulk_email_co_target_id_6cdcd92a52b1f9d9_fk_bulk_email_target_id` FOREIGN KEY (`target_id`) REFERENCES `bulk_email_target` (`id`);

COMMIT;
```

efischer19 commented:
This new approach still adds a M2M relationship with a targets table; however, for the common use `myself`, `staff`, `learners`, and `all` values there will be a single value.

Cohorts will get their own entry on a more specialized table, which can accurately track the cohort to be targeted.

efischer19 commented:
We must move the `CourseEmail` model from storing `to_option` (a choose-from-this-predefined-enum field using a `max_length=64 CharField`) to `targets` (a use-whatever-you-want `ManyToMany` relationship). We're going to need to perform both schema and data migrations. Since the new feature cannot accurately be represented using the current system, there will eventually be a migration which cannot accurately be rolled back without losing data. This plan will hopefully make that switch as safely as possible.

**Proposed Migration Plan** (the safe approach)
- _Release 1_
  - Perform schema migration adding `target` tables to database backend. These will *not* be pre-populated with data.
  - Release code change making API write to both `to_option` and `target` fields.
    - At this point, we will be using code restrictions to ensure that only values which are storable by both systems are being written.
    - We'll also need to ensure that these write are fault-tolerant, and do not report failure unless **both** write attempts fail.
  - Release simultaneous code change making API read from new `target` field first, with a fallback to the old `to_option` field.
    - Note that neither of these code changes will change behavior in any meaningful way.
- __Release 2__
  - Perform data migration, copying any values that are only in the `to_option` field so that they are then stored in both fields.
  - Perform schema migration, dropping the `to_option` column entirely
    - Remember, at this point all entries will exist in the new `target` field, and the code is trying to read from there.
  - Release code change removing dual-read and dual-write functionality
  - Release code change removing restriction on what values may be written
    - This is the first step that is "un-rollback-able". If new data is added into the `targets` field that is not representable by `to_option`, that data will be lost on a rollback.
    - This is the step that will enable new multi-select functionality in production.

efischer19 commented on lms/djangoapps/bulk_email/migrations/0003_add_email_targets.py:
```diff
@@ -0,0 +1,76 @@
+            name='targets',
+            field=models.ManyToManyField(to='bulk_email.Target'),
+        ),
+    ]
```
sqlmigrate data got lost:
```
edxapp@precise64:~/edx-platform$ ./manage.py lms --settings=devstack sqlmigrate bulk_email 0003
BEGIN;
CREATE TABLE `bulk_email_target` (`id` integer AUTO_INCREMENT NOT NULL PRIMARY KEY, `target_type` varchar(64) NOT NULL);
CREATE TABLE `bulk_email_cohorttarget` (`target_ptr_id` integer NOT NULL PRIMARY KEY, `cohort_id` integer NOT NULL);
CREATE TABLE `bulk_email_courseemail_targets` (`id` integer AUTO_INCREMENT NOT NULL PRIMARY KEY, `courseemail_id` integer NOT NULL, `target_id` integer NOT NULL, UNIQUE (`courseemail_id`, `target_id`));
ALTER TABLE `bulk_email_cohorttarget` ADD CONSTRAINT `bulk_emai_target_ptr_id_7974c77c83c2899d_fk_bulk_email_target_id` FOREIGN KEY (`target_ptr_id`) REFERENCES `bulk_email_target` (`id`);
ALTER TABLE `bulk_email_cohorttarget` ADD CONSTRAINT `b_cohort_id_3d66a5e8e283dba0_fk_course_groups_courseusergroup_id` FOREIGN KEY (`cohort_id`) REFERENCES `course_groups_courseusergroup` (`id`);
ALTER TABLE `bulk_email_courseemail_targets` ADD CONSTRAINT `bul_courseemail_id_47818d2b9b38e0e0_fk_bulk_email_courseemail_id` FOREIGN KEY (`courseemail_id`) REFERENCES `bulk_email_courseemail` (`id`);
ALTER TABLE `bulk_email_courseemail_targets` ADD CONSTRAINT `bulk_email_co_target_id_6cdcd92a52b1f9d9_fk_bulk_email_target_id` FOREIGN KEY (`target_id`) REFERENCES `bulk_email_target` (`id`);

COMMIT;
```

efischer19 commented:
@maxrothman Looking at splunk logs of when we hit the `send_email` API endpoint (to create an entry to be picked up by a task later), we're touching this table a few hundred times per day (max of 12 times in a single hour over the past week) at most. Given that, @cahrens and I both think that a short outage, single migration is the way to go, so we'll be proceeding with that.

efischer19 commented:
State of this PR, since it's been a while since i was last here:

- The frontend changes have been moved to https://github.com/edx/edx-platform/pull/12480. It is working to the point of passing new targets into the backend API, but the sent history cannot be updated until the backend exposes that information
- Bulk email is now feature flagged via a configuration model, for ease of deployment.
- Deployment plan:
  - Disable feature using new switch
  - Run 3 migrations, which will be reversible if a quick reset is needed
    - Add new targets table (fast)
    - Move data from old to_option column into new targets table (potentially slow, ought to be tested)
    - Delete old to_option column (potentially slow/dangerous, needs testing)
  - Re-enable feature using switch

I will not be actually enabling cohort targets in this PR, just adding their table as part of the migration. Remaining todo:
- ensure api can handle new target data from frontend
- ensure objects are correctly created with this new paradigm
- expose these fields to the email history section of the frontend
- ensure tasks run with the new targets correctly (potentially separate PR)

efischer19 commented:
@catong @chris-mike I have the initial take of this up and working on my sandbox at https://efischer19.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/instructor#view-send_email. Could you take a look at docstrings/UX respectively?

I've still got some tests to clean up and engineering review to go through, so this is far from final, but I wanted to get you two involved early. Thanks!

catong commented:
@efischer19 Making sure I understand what's on the sandbox - multi-select of checkboxes, but cohort selection not yet available, is that right?

efischer19 commented:
That's correct @catong. There's a bit of backend code for cohorts in place (due to the complication of migrations, I planned all of that work together), but nothing that is visible on the sandbox, that will be following in a separate story.

[This ticket](https://openedx.atlassian.net/browse/TNL-4356)
[The next ticket](https://openedx.atlassian.net/browse/TNL-4357)

catong commented:
Thanks @efischer19. If you don't consider it out of scope, I'd like to suggest edits not to the checkbox labels but to the "Help" text below the message box.

efischer19 commented:
Sure thing @catong! I'm all about the "leave things better than you found them" approach. What do you have in mind?

catong commented:
I'll make inline comments on the lms/templates/instructor/instructor_dashboard_2/send_email.html
file. Just edits for style compliance and hopefully improved clarity.

catong commented on lms/templates/instructor/instructor_dashboard_2/send_email.html:
```diff
@@ -9,38 +9,52 @@
+            %else:
+                <input type="text" id="id_subject" name="subject" maxlength="128" size="75">
+            %endif
+            <span class="tip">${_("(Max 128 characters)")}</span>
```
"Max" -> "Maximum"

catong commented on lms/templates/instructor/instructor_dashboard_2/send_email.html:
```diff
@@ -9,38 +9,52 @@
-  </li>
   </ul>
  <div class="submit-email-action">
   <p class="copy">${_("Please try not to email students more than once per week. Before sending your email, consider:")}</p>
```
@efischer19  I am recommending this text to replace what is currently on lines 60-63, i.e. 2 sentences total, no need for the list items. 
"We recommend sending learners no more than one email message per week. Before you send your email, review the text carefully and send it to yourself first, so that you can preview the formatting and make sure embedded images and links work correctly."

catong replied:
Line 68: "Once the "Send Email" button is clicked..."
-> "When you select **Send Email**, your email message is added to the queue for sending, and cannot be cancelled."

catong replied:
Line 94 : "To see the content of all previously sent..." -> To see the content of previously sent ..."  (delete "all")
Line 100: "To read an email..." -> "To read a sent email message"
Line 107: "To see the status for all bulk email tasks ever submitted..." -> "To see the status for all email tasks submitted for this course..."
Line 107: "...click on this button" -> "click this button" (delete "on")

efischer19 replied:
At line 68, is bolding **Send Email** deliberate?

catong replied:
It needs some formatting to set those words apart as being the UI button label. I'm not sure what is standard in UI text: existing text has single quotation marks - we could stick with those.

efischer19 replied:
I can wrap it in a `<strong>` tag - that ought to allow us to get the style matching the button using css. Thought on this @chris-mike?

<img width="371" alt="screen shot 2016-05-19 at 3 34 48 pm" src="https://cloud.githubusercontent.com/assets/7373924/15407090/46b33620-1dd7-11e6-8915-1a42a89e5b72.png">

catong replied:
I hope @chris-mike has some suggestions for the styling of "CAUTION!"
@efischer19 Sorry if my earlier comment was not clear. The "list-advice" items from lines 62 through 63 should be deleted, as well as the bold and shouty message on line 69 "A queued email CANNOT be cancelled."

catong commented on lms/static/coffee/spec/instructor_dashboard/send_email_spec.coffee:
```diff
@@ -0,0 +1,79 @@
+          params.success()
+        )
+        @send_email.$btn_send.click()
+        expect($('.msg-confirm').text()).toEqual('Your email was successfully queued for sending. Please note that for large classes, it may take up to an hour (or more, if other courses are simultaneously sending email) to send all emails.')
```
"Your email was..." -> "Your email message was..."
"Please note that for..." -> "In courses with a large number of learners, email messages to learners might take up to an hour to be sent."

catong commented on lms/static/coffee/spec/instructor_dashboard/send_email_spec.coffee:
```diff
@@ -0,0 +1,79 @@
+        for target in @send_email.$send_to
+            target.checked = true
+        @send_email.$btn_send.click()
+        expect($('.msg-confirm').text()).toEqual('Your email was successfully queued for sending. Please note that for large classes, it may take up to an hour (or more, if other courses are simultaneously sending email) to send all emails.')
```
Same edits as above at line 58

catong commented on lms/static/coffee/src/instructor_dashboard/send_email.coffee:
```diff
@@ -15,147 +15,154 @@ create_email_content_table = -> window.InstructorDashboard.util.create_email_con
+                # Validation for keyword substitution
+                validation = KeywordValidator().validate_string body
+                if not validation.is_valid
+                    message = gettext("There are invalid keywords in your email. Please check the following keywords and try again:")
```
Remove "please" and replace colon with period: "Check the following keywords and try again."

catong commented on lms/static/coffee/src/instructor_dashboard/send_email.coffee:
```diff
@@ -15,147 +15,154 @@ create_email_content_table = -> window.InstructorDashboard.util.create_email_con
+
+                target_map = {
+                        "myself": gettext("Yourself"),
+                        "staff": gettext("Everyone who is staff or instructor on this course"),
```
"Everyone who is staff or instructor on this course" -> "Everyone who has staff privileges in this course"
@efischer19 Is the reworded message accurate in terms of the privileges?

efischer19 replied:
Yeah, I believe that's accurate.

catong commented on lms/static/coffee/src/instructor_dashboard/send_email.coffee:
```diff
@@ -15,147 +15,154 @@ create_email_content_table = -> window.InstructorDashboard.util.create_email_con
+                target_map = {
+                        "myself": gettext("Yourself"),
+                        "staff": gettext("Everyone who is staff or instructor on this course"),
+                        "learners": gettext("All learners enrolled in this course"),
```
"All learners who are enrolled in this course"

catong commented on lms/static/coffee/src/instructor_dashboard/send_email.coffee:
```diff
@@ -15,147 +15,154 @@ create_email_content_table = -> window.InstructorDashboard.util.create_email_con
+                        "staff": gettext("Everyone who is staff or instructor on this course"),
+                        "learners": gettext("All learners enrolled in this course"),
+                }
+                success_message = gettext("Your email was successfully queued for sending. Please note that for large classes, it may take up to an hour (or more, if other courses are simultaneously sending email) to send all emails.")
```
"Your email was..." -> "Your email message was..."
"Please note that for..." -> "In courses with a large number of learners, email messages to learners might take up to an hour to be sent."

catong commented on lms/static/coffee/src/instructor_dashboard/send_email.coffee:
```diff
@@ -15,147 +15,154 @@ create_email_content_table = -> window.InstructorDashboard.util.create_email_con
+                        "learners": gettext("All learners enrolled in this course"),
+                }
+                success_message = gettext("Your email was successfully queued for sending. Please note that for large classes, it may take up to an hour (or more, if other courses are simultaneously sending email) to send all emails.")
+                confirm_message = gettext("You are about to send an email titled '<%= subject %>' to the following targets. Is this OK?")
```
"You are sending an email message with the subject ```<%=subject %>``` to the following recipients. Is this OK?"

catong commented:
@efischer19 I've finished my editing pass. Thanks!

efischer19 commented on lms/static/coffee/src/instructor_dashboard/util.coffee:
```diff
@@ -1,3 +1,4 @@
+# TODO - lots of safe-commit-linter failures in this file - how can I get HtmlUtils into a coffeescript file?
```
@robrap - I may need your assistance on this one. This is a utility file that I changed ~30 lines in, that is just riddled with `javascript-concat-html` and `javascript-jquery-html` errors.

Is there a simple way to get `HtmlUtils` into this file to fix these issues?

robrap replied:
If it's not using requirejs, you can use `edx.HtmlUtils`.  If that doesn't work, it may be a question for @andy-armstrong.

andy-armstrong replied:
@efischer19 @robrap Yes, you should be able to use ``edx.HtmlUtils`` here.

efischer19 commented on lms/static/coffee/src/instructor_dashboard/util.coffee:
```diff
@@ -206,31 +227,62 @@ create_email_message_views = ($messages_wrapper, emails) ->
+        edx.HtmlUtils.HTML('</em>'),
+      )
+    )
+    $message = $('<div>').html(email_info.email['html_message'])  # safe-lint: disable=javascript-jquery-html
```
The message content of an email is HTML by design, escaping it breaks that.

robrap replied:
I think it is better to use code to declare it, rather than to disable.  I didn't test the following, but I think it would be something like: 
```
$message = edx.HtmlUtils.setHtml(
    $('<div>'),
    edx.HtmlUtils.HTML(email_info.email['html_message'])
)
```

efischer19 replied:
Oh yeah, that ought to work nicely. Cool.

robrap commented on lms/static/coffee/src/instructor_dashboard/util.coffee:
```diff
@@ -206,31 +227,62 @@ create_email_message_views = ($messages_wrapper, emails) ->
+      '<section>',
+      "aria-hidden": "true",
+      class: "modal email-modal",
+      id: edx.HtmlUtils.interpolateHtml("email_message_{id}", {id: email_id})
```
I'm pretty sure that JQuery will be escaping anything before it sets the attributes, so I think this could double-escape.

robrap commented on lms/static/coffee/src/instructor_dashboard/util.coffee:
```diff
@@ -206,31 +227,62 @@ create_email_message_views = ($messages_wrapper, emails) ->
+      type: "button",
+      name: "copy-email-body-text",
+      value: gettext("Copy Email To Editor"),
+      id: edx.HtmlUtils.interpolateHtml("email_message_{id}", {id: email_id})
```
Same deal as above.

robrap commented on lms/static/coffee/src/instructor_dashboard/util.coffee:
```diff
@@ -206,31 +227,62 @@ create_email_message_views = ($messages_wrapper, emails) ->
+    $email_header.append $subject
+    $email_header.append $requester
+    $email_header.append $created
+    $email_header.append $sent_to
```
It's possible that the linter may fail on this in the future because it might be too painful to handle.  Could you switch it to the following while we are here and testing this?:
```
$email_header.append($subject)
```

robrap commented on lms/static/coffee/src/instructor_dashboard/util.coffee:
```diff
@@ -206,31 +227,62 @@ create_email_message_views = ($messages_wrapper, emails) ->
     $email_content = $ '<div>', class: 'email-content-message'
-    $email_content.append $('<h2>', class: "message-bold").html("<em>" + gettext("Message:") + "</em>")
-    $message = $('<div>').html(email_info.email['html_message'])
+    $email_content.append edx.HtmlUtils.setHtml(
```
Same deal with `append(X)` instead of `append X` throughout.  This isn't essential now.  But, I think the linter may require this in the future because the other format is just too difficult to parse.

robrap commented:
Thanks @efischer19.  Looks good except for the minor issues I mentioned.

efischer19 commented:
@robrap I've addressed all your concerns, https://github.com/edx/edx-platform/pull/12301/commits/198f6590b016655a0122f0e571bf7534d8156795#diff-f25e10702aea531db762c68fb4ed7d26 shows all the safe-commit changes in the file you were looking at. I filed https://openedx.atlassian.net/browse/TNL-4633 for the false positives as we discussed.

efischer19 commented on common/djangoapps/util/json_request.py:
```diff
@@ -66,7 +66,7 @@ def __init__(self, resp_obj=None, status=None, encoder=EDXJSONEncoder,
             content = serialize('json', resp_obj)
         else:
-            content = json.dumps(resp_obj, cls=encoder, indent=2, ensure_ascii=False)
+            content = json.dumps(resp_obj, cls=encoder, indent=2, ensure_ascii=True)
```
@mushtaqak This seems to fix the issues you, @robrap, and ! were discussing on your PR earlier: https://github.com/edx/edx-platform/pull/12237/files#r63996545. We're going to have a full test run and see if this is an acceptable default change to make before evaluating our options further.

@nedbat Is there a reason that `ensure_ascii` would need to be False here? It looks like we do this in several places: https://github.com/edx/edx-platform/search?&q=ensure_ascii, but the `dump_js_escaped_json` method uses True already https://github.com/edx/edx-platform/blob/master/openedx/core/djangolib/js_utils.py#L72

robrap replied:
@nedbat: Please see the first link that @efischer19 posted above as well.  It shows how json.dumps() outputs invalid JSON when mixing `ensure_ascii=False` and markupsafe's `Markup`.  Can you see any reason why we couldn't change all of these to `ensure_ascii=True`?

mushtaqak replied:
@efischer19 Thanks, this fixed the issue actually :)

efischer19 replied:
Bump @nedbat, don't want this to slip through the cracks.

Not immediately pressing, but we'd like to get your input here if you get a minute.

nedbat replied:
Using ensure_ascii=True is fine with me everywhere.  Eric and I looked at that other bizarre escaping problem, and it is truly bizarre...

jul replied:
Well, it is not bizarre. It is solving a lot of mysterious bugs in CMS/LMS related to i18n.
- ORA updates impossible with LANG!=en_en
- files names with accents present in module store but export or printing problems in LMS 
....

I found this comment with git blame because I wanted to make sure why this was my fix


I noticed that some of the work was done in async with JSON requests in the CMS. My hypothesis is content are translated using gettext returning unicode strings, while xblock coders codes may have treated strings as strings natively (which they must be naïvely). The ensure asci = True (default value) use a \u escaping and normalizing and always gives keys as str. While (I guess) otherwhise the native type was kept and that a huge lot of encode/decode was made.

I know LTS has this fix, this is just a message to the next coder who will be puzzled to understand why this obscure patch solves l10n problems

efischer19 commented:
@andy-armstrong I'm seeing js test failures due to `TypeError: edx.HtmlUtils is undefined`, but the function seems to be working and escaping things when tested locally. Any advice?
https://build.testeng.edx.org/job/edx-platform-js-pr/19060/consoleFull

andy-armstrong commented:
@efischer19 Unfortunately HtmlUtils isn't defined automatically for all tests. You need to do something like this to ensure that it is defined before the test runs:

https://github.com/edx/edx-platform/blob/master/lms/static/js/spec/main.js#L368

In fact, you had to add this in your last safe template PR (but then the syntax changed with the switch over to Karma):

https://github.com/edx/edx-platform/pull/11886/files

robrap commented:
@efischer19: once you figure out what is needed for the tests, let me know if you think readthedocs requires any helpful updates. Thanks.

chris-mike commented:
@efischer19 My only comment regards the font size in the Pending Task and Task History tables. Can you make them the same size as the paragraph under the "pending tasks" headline, which is (I think) 16px.  Because the information is gibberish (to most people), I'd suggest 12 or 14PX.

efischer19 commented:
Thanks @andy-armstrong that makes sense. I've tried to do that in my latest commit (https://github.com/edx/edx-platform/pull/12301/commits/0473c6af1de6f864385ba2b11740609f2a366854), but for some reason the `init()` function is not being called early enough. Let me know when you get a minute and I'll show you what's going on.

@chris-mike Thanks, I'll let you know once the sandbox is updated.

efischer19 commented:
jenkins run all

efischer19 commented on lms/static/coffee/src/instructor_dashboard/util.coffee:
```diff
@@ -141,18 +140,20 @@ p_wrapper = (value) ->
-unknown_p = p_wrapper(gettext('Unknown'))
+unknown_p = () ->
+  p_wrapper(gettext('Unknown'))
+
```
@andy-armstrong @bjacobel This was what fixed up my last few test failures. `p_wrapper` uses `edx.HtmlUtils`, so by calling that function to define a value, I was hitting the error before `init()` was run to set up the global values. Moving `unknown_p` to an evaluated-when-called function allowed `init()` to do it's thing before HmtlUtils is needed.

efischer19 commented on lms/djangoapps/bulk_email/models.py:
```diff
@@ -83,32 +165,42 @@ def __unicode__(self):
-            fmt = 'Course email being sent to unrecognized to_option: "{to_option}" for "{course}", subject "{subject}"'
-            msg = fmt.format(to_option=to_option, course=course_id, subject=subject)
-            raise ValueError(msg)
+        #from nose.tools import set_trace; set_trace()
```
remove this line

efischer19 commented:
jenkins run js

efischer19 commented:
Huzzah! A green build! I've sliced my commits apart so they hopefully make logical sense for review. The github ordering looks odd, here's the best way to go through them:

1. [Backend model changes](https://github.com/edx/edx-platform/pull/12301/commits/9fd24b11ed0dbcbc5bff8e53f096ff8795815b31)
2. And [the corresponding migrations](https://github.com/edx/edx-platform/pull/12301/commits/24bfda7a6d4579674a27a562638e0fb55c982175)
3. [Python unit test updates](https://github.com/edx/edx-platform/pull/12301/commits/076dfa2182b8ddb57059aa026d82a46b6b245b4c)
4. [Front-end code changes, test additions, and safe template fixes](https://github.com/edx/edx-platform/pull/12301/commits/9312077262dd83755ac6a82278b28fa3d798bbd4) - @cahrens I’ve updated https://github.com/edx/edx-platform/pull/12480 if you just want to look at what’s changed since your last review
5. [Bok choy integration test updates](https://github.com/edx/edx-platform/pull/12301/commits/6e52031409408493d0c94d15955d65d662963e56)
6. [Safe-commit linter false positive cherry-pick](https://github.com/edx/edx-platform/pull/12301/commits/6f38ae692f6d2efcc496d6841a1e1d26b50219f5)
7. [Safe-commit-driven test fix, in need-to-talk-with-Ned status](https://github.com/edx/edx-platform/pull/12301/commits/0315fb22d3ea03265b84ca557b40f52a25d838eb)

**Here's what's remaining before this can be merged:**

- [ ] Code review: @sanfordstudent (Items 1-5 above)
- [ ] Code review: @cahrens (Items 1-5 above, already started 4 previously)
- [ ] Blocking PR to merge: @robrap with https://github.com/edx/edx-platform/pull/12536
- [ ] Minor code review: @nedbat (Item 7 above)
- [ ] Doc review: @catong (check sandbox, I've made the updates we discussed before)
- [ ] UX review: @chris-mike (check sandbox, I've addressed your prior suggestions) 
- [ ] Devops review: @max (Item 2 above, we're planning to run these on the load-test environment to get timing and database load numbers)

The sandbox is currently being updated - https://efischer19.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/instructor#view-send_email with staff@example/com:edx

efischer19 commented:
jenkins run js

cahrens commented on lms/djangoapps/bulk_email/models.py:
```diff
@@ -52,29 +50,113 @@ class Meta(object):
+        if self.target_type == SEND_TO_MYSELF:
+            if user_id is None:
+                raise ValueError("Must define self user to send email to self.")
+            user = User.objects.filter(id=user_id)
```
Nit-- I think you can use "get" if you know there will only be a single matching item.

efischer19 replied:
`get()` will return an object rather than a queryset though, making the combination logic more complicated.

cahrens commented on lms/djangoapps/bulk_email/models.py:
```diff
@@ -52,29 +50,113 @@ class Meta(object):
+        elif self.target_type == SEND_TO_LEARNERS:
+            if staff_instructor_qset.count() <= 0:
+                return use_read_replica_if_available(enrollment_qset)
+            return use_read_replica_if_available(enrollment_qset.exclude(id__in=staff_instructor_qset))
```
Is there any cost in doing the exclude if the staff_instructor_qset is empty? It would simplify the code to just have the "if" clause.

sanfordstudent commented on lms/djangoapps/bulk_email/models.py:
```diff
@@ -52,29 +50,113 @@ class Meta(object):
+                return use_read_replica_if_available(enrollment_qset)
+            return use_read_replica_if_available(enrollment_qset.exclude(id__in=staff_instructor_qset))
+        elif self.target_type == SEND_TO_COHORT:
+            return User.objects.none()  # TODO: cohorts aren't hooked up, put that logic here
```
If cohorts aren't set up yet, why treat this case differently from the default?

cahrens commented on lms/djangoapps/bulk_email/models.py:
```diff
@@ -52,29 +50,113 @@ class Meta(object):
+                use_read_replica_if_available(staff_instructor_qset) |
+                use_read_replica_if_available(enrollment_qset)
+            )
+            return recipient_qsets
```
Why not the read replica in this case?

cahrens commented on lms/djangoapps/bulk_email/models.py:
```diff
@@ -83,32 +165,41 @@ def __unicode__(self):
+            if target not in EMAIL_TARGETS:
+                fmt = 'Course email being sent to unrecognized target: "{target}" for "{course}", subject "{subject}"'
+                msg = fmt.format(target=target, course=course_id, subject=subject)
+                raise ValueError(msg)
```
Will these error messages ever surface to the staff member (including ones above)? If so, they should be marked for translation.

efischer19 replied:
Depends on what you mean by "will". These are backend API-level errors, so our frontend should not be hitting these cases. They're not really intended to be surfaced to the end user; it's more likely they'd be used when an operator is digging through log files to diagnose an unexpected error. Do we traditionally translate such strings?

cahrens commented on lms/djangoapps/bulk_email/tasks.py:
```diff
@@ -220,16 +171,23 @@ def perform_delegate_email_batches(entry_id, course_id, task_input, action_name)
+    combined_set = User.objects.none()
+    for qset in recipient_qsets:
+        combined_set |= qset
+    combined_set = combined_set.distinct()
```
The or operation doesn't make them distinct?

efischer19 replied:
I thought you were right about this, but it turns out the `distinct` is still needed to guarantee no duplicate emails are sent.

Failing test, notice that `robot+test+212@edx.org` is recipient 14 and 15
https://build.testeng.edx.org/job/edx-platform-python-unittests-pr/19262/testReport/lms.djangoapps.bulk_email.tests.test_email/TestEmailSendFromDashboardMockedHtmlToText/test_no_duplicate_emails_staff_instructor/

cahrens commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -3415,7 +3415,7 @@ def setUpClass(cls):
         test_message = u'\u6824 test message'
         cls.full_test_message = {
-            'send_to': 'staff',
+            'send_to': '["myself", "staff"]',
```
Why was myself added?

efischer19 replied:
This `full_test_message` variable is used in several tests, and had to be converted to a single-element array anyways, so I elected to add another value to more adequately test "the API can handle multiple `send_to` values" during those tests.

cahrens commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -2312,7 +2313,7 @@ def list_report_downloads(_request, course_id):
     response_payload = {
         'downloads': [
-            dict(name=name, url=url, link='<a href="{}">{}</a>'.format(url, name))
+            dict(name=name, url=url, link=HTML('<a href="{}">{}</a>').format(HTML(url), name))
```
Should you wrap "name" with Text?

sanfordstudent commented on lms/djangoapps/bulk_email/tests/test_email.py:
```diff
@@ -210,7 +210,7 @@ def test_send_to_all(self):
         test_email = {
             'action': 'Send email',
-            'send_to': 'all',
+            'send_to': '["myself", "staff", "learners"]',
```
I think that if the idea is to send to all except cohort, the name of this test or at least a docstring should reflect that.

efischer19 replied:
It's not intending "all except cohort", just "all users", which requires selecting all options (that are currently enabled)

cahrens commented on lms/djangoapps/bulk_email/models.py:
```diff
@@ -52,29 +50,113 @@ class Meta(object):
+            return use_read_replica_if_available(enrollment_qset.exclude(id__in=staff_instructor_qset))
+        elif self.target_type == SEND_TO_COHORT:
+            return User.objects.none()  # TODO: cohorts aren't hooked up, put that logic here
+        elif self.target_type == SEND_TO_ALL:
```
Do we wish to keep SEND_TO_ALL? If so, some of the test changes below can be simplified.

efischer19 replied:
I'm going to be removing SEND_TO_ALL

sanfordstudent commented on lms/djangoapps/bulk_email/tests/test_models.py:
```diff
@@ -25,9 +25,9 @@ def test_creation(self):
+        email = CourseEmail.create(course_id, sender, [to_option], subject, html_message)
         self.assertEquals(email.course_id, course_id)
-        self.assertEquals(email.to_option, SEND_TO_STAFF)
+        self.assertEquals(email.targets.all()[0].target_type, SEND_TO_STAFF)
```
If it's not relevant for other functionality that `SEND_TO_STAFF` always come first in this list, consider changing this to an `assertTrue` that `targets` contains `SEND_TO_STAFF`.

efischer19 commented on lms/static/coffee/src/instructor_dashboard/send_email.coffee:
```diff
@@ -15,147 +15,154 @@ create_email_content_table = -> window.InstructorDashboard.util.create_email_con
 create_email_message_views = -> window.InstructorDashboard.util.create_email_message_views.apply this, arguments
```
Lots of tab-fixing in this file - to view this file without whitespace changes: https://github.com/edx/edx-platform/pull/12301/files?w=1#diff-84c026f6b037df9822fa7e5caf4ff813

cahrens commented:
There are some problems with the view of the previously sent messages:

![image](https://cloud.githubusercontent.com/assets/484484/15509681/54f9392c-21a2-11e6-9e2c-03b7a69ec457.png)

cahrens commented:
Should copy email to editor (from previously sent list) select who the original recipients were?

@efischer19 I'm done with my pass through. The code is looking great, but I'd recommend more manual testing (in an environment where emails can actually be sent/received).

sanfordstudent commented:
@efischer19 I'm done too. My only comments were minor. Nice!

efischer19 commented:
@cahrens It looks like the copy email to editor functionality has never attempted to copy over the original recipients, but it could easily be added if we desire. I'll bring it up with Shelby at standup tomorrow morning. Original commit: https://github.com/edx/edx-platform/commit/5791fd10871020d35ae29dfcf1bb1f01b1ce9327

The `[object Object]` artifacts were a problem with how I was using `edx.HtmlUtils`, @andy-armstrong got me straightened out there.

efischer19 commented:
jenkins run all

efischer19 commented:
Conditional 👍 from me on the code changes, just a couple small issues to address.

Would like to see the functionality in action before merging this, working with @pdesjardins to make that happen.

efischer19 commented:
jenkins run all

efischer19 commented on lms/static/coffee/src/instructor_dashboard/send_email.coffee:
```diff
@@ -15,147 +15,154 @@ create_email_content_table = -> window.InstructorDashboard.util.create_email_con
+                confirm_message = gettext("You are sending an email message with the subject {subject} to the following recipients. Is this OK?")
+                for target in targets
+                    confirm_message += "\n\t-" + target_map[target]
+                full_confirm_message = edx.StringUtils.interpolate(confirm_message, {subject: subject})
```
@robrap `StringUtils` did the trick here, thanks! 👍 

FYI - 'confirm' is a very low level js call, so I doubt we'll ever be able to modify it for our usage - http://www.w3schools.com/jsref/met_win_confirm.asp

catong commented:
@efischer19 Sorry for the delay reviewing the text on the latest sandbox. The Caution! part of the message below the input box looks really heavy and I don't know what UI patterns we might use: @chris-mike ?

At the least, if Christine does not have any suggestions, can we add an extra line before the "CAUTION!" line and make the last sentence ("A queued email CANNOT be cancelled.") regular weight, not bold?

catong commented:
There's also something odd about the way the subject is displayed in this confirmation message.
![bulkemailconfirmation](https://cloud.githubusercontent.com/assets/9041399/15549684/939900e8-2262-11e6-8ade-95f430219d7a.png)

efischer19 commented:
I'll leave the final call to @chris-mike, but is this what you were thinking @catong? I'll look into the confirmation dialog to see what we can do there.

Current:
<img width="857" alt="screen shot 2016-05-25 at 1 54 17 pm" src="https://cloud.githubusercontent.com/assets/7373924/15550646/72addab6-2280-11e6-902b-0a287b263cb0.png">

Proposed:
<img width="860" alt="screen shot 2016-05-25 at 1 54 35 pm" src="https://cloud.githubusercontent.com/assets/7373924/15550654/7b0fba94-2280-11e6-8f0f-006c34e2a894.png">

catong commented:
Thanks @efischer19 I think the spacing does help. But - oops, I missed one thing - the last sentence actually was one that I had intended to delete in a previous comment. Please delete "A queued email CANNOT be cancelled" since my previous rewrite has already included it in the caution sentence. Sorry about that.

efischer19 commented on lms/static/js/spec/main_requirejs_coffee.js:
```diff
@@ -17,6 +18,22 @@
+                        window.edx.StringUtils = StringUtils;
+                    });
+                }
+            }
```
Hey @andy-armstrong, I'm running into some more troubles getting `edx.StringUtils` loaded properly in tests. `coffee/src/instructor_dashboard/send_email.coffee` is run with `lms-coffee`, so the previous use of `js/spec/main.js` didn't help. Is this the right file to be making this modification for `lms-coffee` tests? It doesn't seem to be working for me.

sanfordstudent commented:
Since I'll be out of town for the conclusion of this code review, 👍  for the portion I reviewed.

efischer19 commented:
@catong The sandbox is updating again, I've tried to implement your suggestions. Let me know what you think!

catong commented:
@efischer19 Thanks, looks good 👍

efischer19 commented:
jenkins run bokchoy

efischer19 commented:
Summary of my migration investigation with @feanil:
- On load test, with ~3k rows, the data migration took 23 seconds
- Prod has ~30k rows, so we're looking at about 4 minutes to run everything
- Dropping columns is bad, I'll be removing the 0006 migration from this PR

Here's some sql illustrating what will be run: https://gist.github.com/efischer19/8cf79609b624bc2c20ca7ac4392c8685

So, given those constraints, the plan is:
- flip the admin switch to disable bulk_email. This will prevent new additions being made to the old table during the migration.
- run migrations, copy all data to the new table
- deploy new code
- flip the admin switch to re-enable the feature, begin using the new table for read and write

efischer19 commented:
@cahrens This PR is ready for re-review, everything that's changed since you last viewed is in the "review response" commit.

Sandbox has been updated as well.

cahrens commented on lms/static/coffee/src/instructor_dashboard/util.coffee:
```diff
@@ -240,10 +237,7 @@ create_email_message_views = ($messages_wrapper, emails) ->
-      $close_button,
-      $('<i>', class: 'icon fa fa-times')
-    )
+    $close_button.append($('<i>', class: 'icon fa fa-times'))
```
@ssemenova I'm not sure if you found this i tag usage in your cleanup, but it will need to be fixed.

cahrens commented:
👍

efischer19 commented:
jenkins run bokchoy
